File: frontend-mobile/lib/services/Cognitive_Progress_Prediction/cognitive_progress_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ProgressPredictionApi {
  final String baseUrl;

  ProgressPredictionApi({required this.baseUrl});

  /// ✅ Predict progress using ML service
  /// POST /chromabloom/cognitiveProgress_2/predict-progress
  Future<Map<String, dynamic>> predictProgress(
    Map<String, dynamic> features, {
    int topK = 10,
  }) async {
    final url =
        Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2/predict-progress');

    final res = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        "features": features,
        "top_k": topK,
      }),
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Prediction failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Save a prediction record to MongoDB
  /// POST /chromabloom/cognitiveProgress_2
  Future<Map<String, dynamic>> savePrediction({
    required String userId,
    required double progressPrediction,
    List<Map<String, dynamic>>? positiveFactors,
    List<Map<String, dynamic>>? negativeFactors,
  }) async {
    final url = Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2');

    final res = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        "userId": userId,
        "progress_prediction": progressPrediction,
        "positive_factors": positiveFactors ?? [],
        "negative_factors": negativeFactors ?? [],
      }),
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Save failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Fetch all progress records
  /// GET /chromabloom/cognitiveProgress_2
  Future<Map<String, dynamic>> getAllProgress() async {
    final url = Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2');

    final res = await http.get(
      url,
      headers: {'Content-Type': 'application/json'},
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Fetch all failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Fetch saved predictions by userId (childId)
  /// GET /chromabloom/cognitiveProgress_2/user/:userId
  Future<Map<String, dynamic>> getPredictionsByUserId(String userId) async {
    final url =
        Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2/user/$userId');

    final res = await http.get(
      url,
      headers: {'Content-Type': 'application/json'},
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Fetch failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Fetch a single progress record by record ID
  /// GET /chromabloom/cognitiveProgress_2/:id
  Future<Map<String, dynamic>> getProgressById(String id) async {
    final url = Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2/$id');

    final res = await http.get(
      url,
      headers: {'Content-Type': 'application/json'},
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Fetch by ID failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Update a progress record
  /// PUT /chromabloom/cognitiveProgress_2/:id
  Future<Map<String, dynamic>> updateProgress(
    String id, {
    String? userId,
    double? progressPrediction,
    List<Map<String, dynamic>>? positiveFactors,
    List<Map<String, dynamic>>? negativeFactors,
  }) async {
    final url = Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2/$id');

    final Map<String, dynamic> body = {};
    if (userId != null) body['userId'] = userId;
    if (progressPrediction != null) body['progress_prediction'] = progressPrediction;
    if (positiveFactors != null) body['positive_factors'] = positiveFactors;
    if (negativeFactors != null) body['negative_factors'] = negativeFactors;

    final res = await http.put(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(body),
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Update failed: ${res.statusCode} ${res.body}');
    }
  }

  /// ✅ Delete a progress record
  /// DELETE /chromabloom/cognitiveProgress_2/:id
  Future<Map<String, dynamic>> deleteProgress(String id) async {
    final url = Uri.parse('$baseUrl/chromabloom/cognitiveProgress_2/$id');

    final res = await http.delete(
      url,
      headers: {'Content-Type': 'application/json'},
    );

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else {
      throw Exception('Delete failed: ${res.statusCode} ${res.body}');
    }
  }
}

File: frontend-web/src/services/Therapist/cognitiveProgressService.js
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000';

const cognitiveProgressApi = axios.create({
  baseURL: `${API_BASE_URL}/chromabloom/cognitiveProgress_2`,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 10000,
});

// Request interceptor for auth tokens (if needed)
cognitiveProgressApi.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('authToken');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Response interceptor for error handling
cognitiveProgressApi.interceptors.response.use(
  (response) => response,
  (error) => {
    const customError = {
      message: error.response?.data?.message || error.message,
      status: error.response?.status,
      data: error.response?.data,
    };
    return Promise.reject(customError);
  }
);

/**
 * Get cognitive progress predictions by user ID
 * @param {string} userId - The user/child ID to fetch progress for
 * @returns {Promise<{success: boolean, count: number, data: Array}>}
 */
export const getProgressByUserId = async (userId) => {
  if (!userId) {
    throw new Error('userId is required');
  }

  const response = await cognitiveProgressApi.get(`/user/${userId}`);
  return response.data;
};

/**
 * Get all cognitive progress records
 * @returns {Promise<{success: boolean, count: number, data: Array}>}
 */
export const getAllProgress = async () => {
  const response = await cognitiveProgressApi.get('/');
  return response.data;
};

/**
 * Get single progress record by ID
 * @param {string} id - The progress record ID
 * @returns {Promise<{success: boolean, data: Object}>}
 */
export const getProgressById = async (id) => {
  const response = await cognitiveProgressApi.get(`/${id}`);
  return response.data;
};

/**
 * Create new progress prediction
 * @param {Object} payload - { userId: string, progress_prediction: number }
 * @returns {Promise<{success: boolean, data: Object}>}
 */
export const createProgress = async (payload) => {
  const response = await cognitiveProgressApi.post('/', payload);
  return response.data;
};

/**
 * Update progress prediction
 * @param {string} id - The progress record ID
 * @param {Object} payload - { userId?: string, progress_prediction?: number }
 * @returns {Promise<{success: boolean, data: Object}>}
 */
export const updateProgress = async (id, payload) => {
  const response = await cognitiveProgressApi.put(`/${id}`, payload);
  return response.data;
};

/**
 * Delete progress prediction
 * @param {string} id - The progress record ID
 * @returns {Promise<{success: boolean, message: string, data: Object}>}
 */
export const deleteProgress = async (id) => {
  const response = await cognitiveProgressApi.delete(`/${id}`);
  return response.data;
};

/**
 * Get prediction from Python ML service (via Node proxy)
 * @param {Object} features - Required feature object
 * @param {number} [top_k=10] - Number of top factors to return
 * @returns {Promise<{message: string, result: Object}>}
 */
export const predictProgress = async (features, top_k = 10) => {
  const response = await cognitiveProgressApi.post('/predict-progress', {
    features,
    top_k,
  });
  return response.data;
};

export default {
  getProgressByUserId,
  getAllProgress,
  getProgressById,
  createProgress,
  updateProgress,
  deleteProgress,
  predictProgress,
};